### PR TITLE
fix(vonage): centralize configuration load and add unit tests for tim…

### PR DESCRIPTION
**Pull request by theainerd**
…eouts

## Summary by CodeRabbit

* **Chores**
  * Improved Vonage API timeout configuration management through centralized settings with consistent default values to enhance system reliability and maintainability.

* **Tests**
  * Added comprehensive test coverage for Vonage websocket connection handling, including validation of successful connections, timeout scenarios, and error recovery paths.

### CODE REVIEW

**Review by a4i-architect — COMMENTED** [review]
### High‑Level Summary
This PR moves Vonage timeout configuration from hardcoded env lookups into the central `Settings` object and adds a comprehensive test suite. Centralising config is aligned with standards, but the tests introduce several issues (logging, secret handling, architecture) and there is one architectural concern in the service file related to repeated settings instantiation.

Below are all actionable issues confined strictly to added/modified lines.

---

## Issues

### 1. Repeated settings instantiation (MEDIUM)
**File:** `ConferenceV2/app/services/communication_api/vonage_api.py`  
**Lines:** the two modified timeout assignments using `get_settings()`  
**Issue:** Calling `get_settings()` twice at import-time can instantiate multiple `Settings` objects. This violates the *architecture* guideline (config centralisation and avoiding scattered initialisation).  
**Fix:** Assign a single local `_settings = get_settings()` once, then use its fields.

---

### 2. Direct patching of module‑level config constants (LOW)
**File:** `ConferenceV2/tests/test_vonage_api_timeout.py`  
**Multiple lines:** patching `_VONAGE_GET_CALL_TIMEOUT_SECONDS` and `_VONAGE_UPDATE_CALL_TIMEOUT_SECONDS`  
**Issue:** Tests override module-level constants, which is brittle. Not a functional bug, but violates maintainability expectations.  
**Fix:** Patch the settings object or dependency‑inject timeout values into `DummyVonage`, rather than mocking internal module constants.

---

### 3. Missing structured logging in test-created subclass (LOW)
**File:** `ConferenceV2/tests/test_vonage_api_timeout.py`  
**Function:** `DummyVonage.__init__`  
**Issue:** The subclass omits initialising the real logger, which can cause missing context logs during error propagation. This violates the global *logging* guideline.  
**Fix:** Assign `self.logger = logger_instance` or patch it to a structured logger mock.

---

### 4. Direct use of `time.sleep()` in patched functions (MEDIUM)
**File:** `ConferenceV2/tests/test_vonage_api_timeout.py`  
**Functions:** `slow_get_call`, `slow_update_call`  
**Issue:** Using `time.sleep()` inside async test scopes can block the event loop, violating the async consistency rule.  
**Fix:** Use `asyncio.sleep()` and make those side‑effect functions async-compatible, or patch using `side_effect=asyncio.sleep`.

---

### 5. Potential accidental env leakage due to `sys.path` manipulation (LOW)
**File:** `ConferenceV2/tests/test_vonage_api_timeout.py`  
**Line:** added `sys.path.append(...)`  
**Issue:** Manipulating `sys.path` at runtime can import unintended modules when running under different environments, increasing test fragility.  
**Fix:** Use project‑root‑anchored pytest configuration (`pytest.ini`, `PYTHONPATH`) instead of runtime mutation.

---

## Score Calculation
- MEDIUM: 2 issues = -4  
- LOW: 3 issues = -3  
Total deduction: 7 points  
**Final Score: 93**

---

No references required.

**Review by a4i-architect — COMMENTED** [review]
### High‑Level Summary
This PR consolidates two Vonage timeout environment variables into one and adds a full test suite for websocket connection timeout behavior. Most changes are positive, but there are a few correctness issues, plus one logging violation of team standards.

---

## Issues

### 1. **Leaking Sensitive Data in Logs**  
**File:** `vonage_api.py`  
**Function:** `_try_connecting_websocket_with_participant`  
**Severity:** HIGH (-3)

The modified log lines interpolate `participant.phone_number`, which violates the global logging guideline forbidding PII in logs.

**Fix:**  
Mask the phone number before logging, e.g. introduce a helper that returns only last 2 digits. Replace the PII in all newly added/modified log lines in this PR.

---

### 2. **Magic Timeout Value Not Moved to Config**  
**File:** `vonage_api.py`  
**Global-level variable:** `_VONAGE_CALL_TIMEOUT_SECONDS`  
**Severity:** MEDIUM (-2)

The PR consolidates two previous timeouts into a new one but still defines it inline using `os.environ.get`. This violates the architecture rule requiring all configuration to be centralized in `config.py`.

**Fix:**  
Define `VONAGE_CALL_TIMEOUT_SECONDS` in `Settings` in `config.py` and import it where needed.

---

### 3. **Missing Structured Logging**  
**File:** `vonage_api.py`  
**Function:** `_try_connecting_websocket_with_participant`  
**Severity:** LOW (-1)

The updated log lines still rely on string formatting. The team standard requires structured logging. This is a pre-existing issue, but the PR touches these lines and continues the pattern.

**Fix:**  
Use structured logging, e.g., `logger_instance.warning(message, extra={ ... })`.

---

### 4. **Test Imports Bypass Project Structure**  
**File:** `test_vonage_api_timeout.py`  
**Top-level import hack using sys.path.append**  
**Severity:** LOW (-1)

Modifying `sys.path` in tests is an antipattern and complicates maintainability.

**Fix:**  
Use proper pytest path configuration or a local conftest for path resolution.

---

## Score Calculation
- HIGH: 1 issue = -3  
- MEDIUM: 1 issue = -2  
- LOW: 2 issues = -2  
**Total deduction: 7**

**Final Score = 100 - 7 = 93**

---

No references required.

File: ConferenceV2/tests/test_vonage_api_timeout.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        self.ws_server_url = "ws://localhost:8000"
+        self.conf_id = "test-conf"
+        self.participant_info_map = {}
+        self.vonage_conv_id = None
+        self.is_websocket_connected = False
+
+
+@pytest.mark.asyncio
+async def test_try_connecting_websocket_success():
+    api = DummyVonage()
```
Comment (farmanahmed888) [inline ConferenceV2/tests/test_vonage_api_timeout.py:25]: here we are testing dummy Vonage not the real Vonage

File: ConferenceV2/app/services/communication_api/vonage_api.py [hunk 2]
```diff
@@ -21,11 +21,8 @@
 # can block the FastAPI event loop indefinitely (observed: 15+ minute hang on
 # GET /v1/calls/<uuid>). Used by _try_connecting_websocket_with_participant
 # below to bound the sync calls that live on the websocket-attach hot path.
-_VONAGE_GET_CALL_TIMEOUT_SECONDS = float(
-    os.environ.get("VONAGE_GET_CALL_TIMEOUT_SECONDS", "30.0")
-)
-_VONAGE_UPDATE_CALL_TIMEOUT_SECONDS = float(
-    os.environ.get("VONAGE_UPDATE_CALL_TIMEOUT_SECONDS", "30.0")
+_VONAGE_CALL_TIMEOUT_SECONDS = float(
+    os.environ.get("VONAGE_CALL_TIMEOUT_SECONDS", "30.0")
```
Comment (farmanahmed888) [inline ConferenceV2/app/services/communication_api/vonage_api.py]: this was not required it should be
`_VONAGE_CALL_TIMEOUT_SECONDS=30.0`

**Review by a4i-architect — COMMENTED** [review]
### High‑Level Summary
Centralising Vonage timeout configuration is a positive change, but the PR introduces one **broken test**, one **missing error‑logging violation**, and one **config consistency gap**. These issues are small but directly actionable.

---

### Issues

#### 1. Missing error logging context (MEDIUM, -2)
**File:** `vonage_api.py`  
**Function:** `_try_connecting_websocket_with_participant`  
**Issue:** The modified timeout warnings now interpolate a new setting but still do not include full context required by the **Global ERROR HANDLING Standard** (function name + input summary).  
Because the PR modified these lines, the missing compliance becomes relevant here.

**Fix:** Update each warning invocation added/modified in this PR to prepend function name and partial participant info (masked phone number). Example: log `"func=_try_connecting_websocket_with_participant participant=***123 id=<id> ..."`. Do not log raw phone numbers.

---

#### 2. Sensitive data exposed in logs (HIGH, -3)
**File:** `vonage_api.py`  
**Function:** `_try_connecting_websocket_with_participant`  
**Issue:** The modified log lines still include the full phone number, violating **LOGGING → Sensitive data must never appear**.  
Since these lines are changed in this PR (they interpolate the new `_VONAGE_CALL_TIMEOUT_SECONDS`), the issue becomes in‑scope.

**Fix:** Mask phone numbers before logging. Example: `participant.phone_number[-4:]` or a helper such as `mask_phone(phone)`.

---

#### 3. Inadequate config typing (LOW, -1)
**File:** `config.py`  
**Issue:** `VONAGE_CALL_TIMEOUT_SECONDS` is typed as `float`. According to **STRICT TYPING**, DTOs must be strictly typed, and environment-loaded floats are fragile. For seconds values, `int` is more predictable and prevents fractional string parsing edge cases.

**Fix:** Change the setting to `int` and convert fractions in the loader only if needed.

---

#### 4. Invalid test patch path breaks test execution (HIGH, -3)
**File:** `test_vonage_api_timeout.py`  
**Functions:** all tests using `patch("app.services.communication_api.vonage_api._VONAGE_CALL_TIMEOUT_SECONDS", ...)`  
**Issue:** `_VONAGE_CALL_TIMEOUT_SECONDS` is now defined at import time via `get_settings()`. It is no longer a mutable module-level constant tied to environment variables; patching it this way will not work reliably across test environments and depending on import order may patch nothing.

This will lead to broken or flaky tests.

**Fix:**  
Patch `get_settings().VONAGE_CALL_TIMEOUT_SECONDS` instead, e.g.  
`with patch("config.get_settings") as mock_settings:` and set `.VONAGE_CALL_TIMEOUT_SECONDS`.

---

### Score Calculation
Issues:  
- HIGH: 2 × -3 = -6  
- MEDIUM: 1 × -2 = -2  
- LOW: 1 × -1 = -1  

**Total deductions: 9**

**Final Score = 100 − 9 = 91**

---

### Optional References
- [Pydantic Settings Management](https://docs.pydantic.dev/latest/concepts/pydantic_settings/)  
- [OWASP Logging Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Logging_Cheat_Sheet.html)

File: ConferenceV2/app/services/communication_api/vonage_api.py (hunk 2 above)
Comment (theainerd) [reply ConferenceV2/app/services/communication_api/vonage_api.py]: done

**Comment by theainerd** [comment]
testing results on real endpoints.

[Vonage_API_Endpoints_400_404_Verification_Report.docx](https://github.com/user-attachments/files/28552846/Vonage_API_Endpoints_400_404_Verification_Report.docx)

File: ConferenceV2/tests/test_vonage_api_timeout.py (hunk 1 above)
Comment (theainerd) [reply ConferenceV2/tests/test_vonage_api_timeout.py:25]: https://github.com/A4i-tech/SEEDS/pull/215#issuecomment-4612703493

File: ConferenceV2/config.py [hunk 3]
```diff
@@ -81,6 +81,8 @@ class Settings(BaseSettings):
     AUDIO_CAPTURE_CHANNELS: int = 1
     AUDIO_CAPTURE_SAMPLE_WIDTH_BYTES: int = 2
 
+    VONAGE_CALL_TIMEOUT_SECONDS: float = 30.0
```
Comment (N1kunj1998) [inline ConferenceV2/config.py:84]: ## Risk: Old Env Var Names Silently Ignored (Breaking Change)

### What changed

PR replaces two separate env vars with one:

| Removed | Replacement |
|---|---|
| `VONAGE_GET_CALL_TIMEOUT_SECONDS` | `VONAGE_CALL_TIMEOUT_SECONDS` |
| `VONAGE_UPDATE_CALL_TIMEOUT_SECONDS` | `VONAGE_CALL_TIMEOUT_SECONDS` |

### The risk

Any deployment that has `VONAGE_GET_CALL_TIMEOUT_SECONDS` or `VONAGE_UPDATE_CALL_TIMEOUT_SECONDS` set in `.env` / environment will **silently fall back to 30.0** with no error, no warning, no log line. Ops team has no way to know their configured value is being ignored.

Verified locally — setting old vars to `99` and `88` produces `30.0` with zero indication.

### Fix

Add a startup warning in `config.py`:

```python
# config.py
from pydantic import validator
import os, warnings

class Settings(BaseSettings):
    VONAGE_CALL_TIMEOUT_SECONDS: float = 30.0

    @validator("VONAGE_CALL_TIMEOUT_SECONDS", pre=True, always=True)
    def _warn_legacy_vonage_timeout_vars(cls, v):
        for old in ("VONAGE_GET_CALL_TIMEOUT_SECONDS", "VONAGE_UPDATE_CALL_TIMEOUT_SECONDS"):
            if os.environ.get(old):
                warnings.warn(
                    f"{old} is no longer used. Set VONAGE_CALL_TIMEOUT_SECONDS instead.",
                    DeprecationWarning,
                    stacklevel=2,
                )
        return v
```

This way any deployment with stale env vars gets a visible `DeprecationWarning` at startup instead of silent misconfiguration.

**Review by N1kunj1998 — COMMENTED** [review]
### 1. Unit Tests — 6/6 PASS ✅

```
poetry run pytest tests/test_vonage_api_timeout.py -v
```

```
test_try_connecting_websocket_success               PASSED
test_try_connecting_websocket_not_answered          PASSED
test_try_connecting_websocket_get_call_timeout      PASSED
test_try_connecting_websocket_get_call_sdk_error    PASSED
test_try_connecting_websocket_update_call_timeout   PASSED
test_try_connecting_websocket_update_call_sdk_error PASSED
6 passed in 46.99s
```

> ⚠️ Suite takes **47s** because 2 tests use real `time.sleep(2.0)` in background threads. Slow CI + leaked threads. Should be mocked (see risk comment).

---

### 2. Config Wiring ✅

| Scenario | Expected | Got |
|---|---|---|
| Default value | `30.0` | `30.0` ✅ |
| Old vars `VONAGE_GET_CALL_TIMEOUT_SECONDS=99` + `VONAGE_UPDATE_CALL_TIMEOUT_SECONDS=88` | `30.0` (ignored) | `30.0` ✅ |
| `VONAGE_CALL_TIMEOUT_SECONDS=5.0` | `5.0` | `5.0` ✅ |
| Module constant when env set before process start | correct | ✅ |

New env var correctly read by `Settings`. Old env var names have zero effect (silently ignored — see risk comment).

---

### 3. Timeout Mechanism ✅

| Run | Scenario | Result |
|---|---|---|
| A | Real Vonage 404, threshold 30s | `False` in 0.677s ✅ |
| B | Real in-flight request, threshold 0.001s | `asyncio.TimeoutError` in 0.033s ✅ |
| C | Hung leg 30s, threshold 3s | Cut at 3.004s ✅ |

Timeout fires against real in-flight HTTP — not just mocks. `update_call` timeout covered by unit test but not separately probed live.

---

**Review by N1kunj1998 — APPROVED** [review]
LGTM
